### PR TITLE
Support filter syntax

### LINKED ISSUE #28: Support filter syntax in expressions

### PR DESCRIPTION
**Pull request by rictic**
Fixes #28

### CODE REVIEW

File: lib/upgrade_html.js [hunk 1]
```diff
@@ -313,148 +309,108 @@ class Page {
...
 
+    let addNewDeclaration = (declaration, expression, node) => {
+      if (!declaration) {
+        return;
+      }
+      if (newDeclarations) {
+        newDeclarations.push(declaration);
+      } else {
+        this.insertHtmlCommentBefore(node, [
+          'The expression {{' + expression + '}} can\'t work in a',
```
Comment (azakus) [inline lib/upgrade_html.js]: backticks

File: lib/upgrade_html.js [hunk 2]
```diff
@@ -313,148 +309,108 @@ class Page {
...
           attrName += '$';
         }
         if (newDeclarations) {
-          if (newExpression != expression) {
-            node.attribs[attrName] = '{{' + newExpression + '}}';
-          }
-          if (newDeclaration) {
-            newDeclarations.push(newDeclaration);
-          }
+          node.attribs[attrName] = '{{' + newExpression + '}}';
```
Comment (azakus) [inline lib/upgrade_html.js]: backticks

File: lib/upgrade_html.js [hunk 3]
```diff
@@ -313,148 +309,108 @@ class Page {
...
-          if (newDeclaration) {
-            this.insertHtmlCommentBefore(node, [
-              'This expression can\'t work in a dom-bind template, as it should',
-              'be an anonymous computed property. If you convert it into a',
-              'Polymer element then polyup should be able to upgrade it.'
-            ]);
-          } else if (newExpression != expression) {
-            node.attribs[attrName] = '{{' + newExpression + '}}';
-          }
+          node.attribs[attrName] = '{{' + expression + '}}';
```
Comment (azakus) [inline lib/upgrade_html.js]: backticks

File: lib/upgrade_html.js [hunk 4]
```diff
@@ -313,148 +309,108 @@ class Page {
...
-          this.insertHtmlCommentBefore(expressionTextNode, [
-            'This expression can\'t work in a dom-bind template, as it should',
-            'be an anonymous computed property. If you convert it into a',
-            'Polymer element then polyup should be able to upgrade it.'
-          ]);
-          expressionTextNode.data = '{{' + expression + '}}';
+      }
+      // Expression takes up entire text node.
+      if (pieces.length === 1 && ('expression' in pieces[0])) {
+        textNode.data = '{{' + pieces[0].expression + '}}';
```
Comment (azakus) [inline lib/upgrade_html.js]: backticks

File: lib/upgrade_html.js [hunk 5]
```diff
@@ -313,148 +309,108 @@ class Page {
...
+        continue;
+      }
+      for (let piece of pieces) {
+        if ('expression' in piece) {
+          this.$(textNode).before(
+              this.$('<span>').text('{{' + piece.expression + '}}'));
```
Comment (azakus) [inline lib/upgrade_html.js]: backticks

File: lib/upgrade_html.js [hunk 6]
```diff
@@ -476,6 +432,46 @@ class Page {
     });
   }
 
+  matchExpressions(str) {
+    function matchFirstExpression(str) {
+      let match = str.match(/\{\{(.+?)\}\}/);
```
Comment (azakus) [inline lib/upgrade_html.js]: also square braces

File: lib/upgrade_html.js [hunk 7]
```diff
@@ -476,6 +432,46 @@ class Page {
...
+      }
+      let expressionPieces = match[1].split(/\s+\|\s+/);
+      let expression = expressionPieces[0];
+      for (let func of expressionPieces.slice(1)) {
+        let parsed = espree.parse(func).body[0].expression;
+        if (parsed.type === 'CallExpression') {
+          parsed.arguments.unshift(espree.parse(expression).body[0].expression);
+          expression = escodegen.generate(parsed);
+        } else {
+          expression = func + '(' + expression + ')';
```
Comment (azakus) [inline lib/upgrade_html.js]: backticks

**Comment by azakus** [comment]
LGTM with nits

File: lib/upgrade_html.js (hunk 1 above)
Comment (rictic) [inline lib/upgrade_html.js]: Done

File: lib/upgrade_html.js (hunk 2 above)
Comment (rictic) [inline lib/upgrade_html.js]: Done

File: lib/upgrade_html.js (hunk 3 above)
Comment (rictic) [inline lib/upgrade_html.js]: Done

File: lib/upgrade_html.js (hunk 4 above)
Comment (rictic) [inline lib/upgrade_html.js]: Done

File: lib/upgrade_html.js (hunk 5 above)
Comment (rictic) [inline lib/upgrade_html.js]: Done

File: lib/upgrade_html.js (hunk 6 above)
Comment (rictic) [inline lib/upgrade_html.js]: Punting to #40

File: lib/upgrade_html.js (hunk 7 above)
Comment (rictic) [inline lib/upgrade_html.js]: Done

**Pull request merged**